enhancement(http): log source IP as `host` key in the `http` source

NishantJoshi00 opened the pull request:
closes #4221

NishantJoshi00 commented:
Hey @jszwedko & @neuronull , any update on this PR?

neuronull commented on src/sources/util/http/prelude.rs:
```diff
@@ -186,6 +190,7 @@ pub trait HttpSource: Clone + Send + Sync + 'static {
 
             let span = Span::current();
             let make_svc = make_service_fn(move |conn: &MaybeTlsIncomingStream<TcpStream>| {
+                let remote_addr = conn.peer_addr();
```
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏I'd suggest either removing this var and calling it directly in the `PeerAddr` creation, or cloning this var on line `200`.

neuronull commented on src/sources/http_server.rs:
```diff
@@ -484,6 +485,16 @@ impl HttpSource for SimpleHttpSource {
                     );
+
+                    source_ip.map(|addr| {
+                        self.log_namespace.insert_source_metadata(
+                            SimpleHttpConfig::NAME,
+                            log,
+                            Some(LegacyKey::Overwrite(path!("host"))),
+                            path!("host"),
+                            addr.to_string(),
```
I think this necessitates an update to the `schema_definition()` function.

Some background context in https://github.com/vectordotdev/vector/blob/master/docs/tutorials/lognamespacing.md

This should also (hopefully) prompt a unit test change as well.

NishantJoshi00 replied:
Hey @neuronull, I had one more concern regarding using `"host"` instead of `host_key`. Should, I use host key as they key for this instead of "host"?

neuronull replied:
Ah, I believe "host" is the correct usage, as you have here.

neuronull commented on src/sources/util/http/prelude.rs:
```diff
@@ -218,6 +227,10 @@ pub trait HttpSource: Clone + Send + Sync + 'static {
 }
 
+#[derive(Clone)]
+#[repr(transparent)]
```
I'm not super familiar with this macro, can you expand on why it's needed?

NishantJoshi00 replied:
It's just a minor memory optimization. Basically, this suggests the compiler that the type itself doesn't have any size. So, while performing compiler optimization this types gets abstracted as the inner type.

NishantJoshi00 replied:
Here is a reference if you wish to read about it: https://doc.rust-lang.org/nomicon/other-reprs.html#reprtransparent

neuronull reviewed (commented):
Thanks for this contribution.

Overall looks good, have a couple thoughts.

neuronull commented on .test.output:
```diff
@@ -0,0 +1 @@
+cargo nextest run --workspace --no-fail-fast --no-default-features --features "default" 
```
<!-- Warning represent user-facing problems. If possible, it's great to follow this kind of comment with a suggestion.  -->
**🔨 warning:** ‏. 👀

neuronull commented on src/sources/http_server.rs:
```diff
@@ -133,6 +133,12 @@ pub struct SimpleHttpConfig {
     path_key: OptionalValuePath,
 
+
+    /// The event key in which the remote ip is stored.
```
Since we are using the log_schema as the default for this one, I think this comment should follow what we do in other components that do this. There are a few but this is one example: https://github.com/vectordotdev/vector/blob/56486bafe6ce41a7c92a11ccd0e2cf6e8f7ef838/src/sources/docker_logs/mod.rs#L68-L72 (ofc, mentioning the remote IP though)

neuronull reviewed (commented):
Coming along nicely, thanks for making suggested changes.

neuronull commented on src/sources/http_server.rs:
```diff
@@ -133,6 +133,12 @@ pub struct SimpleHttpConfig {
     path_key: OptionalValuePath,
 
+
+    /// The event key in which the remote ip is stored.
```
Note that you'll need to regenerate the CUE doc files for the website documentation rendering.

Run:

```
$ make generate-component-docs
```

, committing the changes.

NishantJoshi00 replied:
Thanks! will do that.

NishantJoshi00 replied:
Hey, I noticed something. the command description for `generate-component-docs` is:
```
generate-component-docs: ## Generate per-component Cue docs from the configuration schema.
	${MAYBE_ENVIRONMENT_EXEC} cargo build $(if $(findstring true,$(CI)),--quiet,)
	target/debug/vector generate-schema > /tmp/vector-config-schema.json 2>/dev/null
	${MAYBE_ENVIRONMENT_EXEC} cargo vdev build component-docs /tmp/vector-config-schema.json \
		$(if $(findstring true,$(CI)),>/dev/null,)
```

what if I am using a shared target directory? I faced this issue, as my build artifacts were stored in a different directory than `target/debug`. I don't think anyone will face this issue, but just wanted to let you know.

NishantJoshi00 replied:
I am also facing issue while running the `cue` command.

```
in `<main>': undefined method `filter_map' for #<Hash:0x0000000133827d58> (NoMethodError)
Did you mean?  filter
```

NishantJoshi00 replied:
> Author

This is resolved, I unset my common target, but facing issues with `cue`. Can you tell me which stable version of cue you folks are using?

NishantJoshi00 replied:
> Note that you'll need to regenerate the CUE doc files for the website documentation rendering.
> 
> Run:
> 
> ```
> $ make generate-component-docs
> ```
> 
> , committing the changes.

This is resolved, I unset my common target, but facing issues with cue. Can you tell me which stable version of cue you folks are using?

neuronull replied:
> what if I am using a shared target directory? I faced this issue, as my build artifacts were stored in a different directory than target/debug. I don't think anyone will face this issue, but just wanted to let you know

Ah good observation. I don't recall this coming up before. I will open a tracking issue.

https://github.com/vectordotdev/vector/issues/19843

neuronull replied:
> Can you tell me which stable version of cue you folks are using?

Ah. I will regenerate the docs for you here. If you don't already have CUE installed. I will spare you this misery.

If you continue to encounter the need for this in future PRs, I believe we are using `v0.4.2`.

neuronull replied:
Actually, I ran into some weirdness trying to push to your branch. If you don't mind you can try installing CUE.

If you run into issues, I can come back to this 🙏

NishantJoshi00 replied:
:crying_cat_face: As all issues go, its a simple issue of it works on my machine, didn't work on my macbook as ruby version 2.* doesn't support the functionality of `filter_map`, tried on my Linux box, worked like a charm. 

And some how, macos isn't easy enough to allow me to update the version of the ruby that is currently installed.

Any who, I have done the required changes, please let me know, if anything is left.

neuronull replied:
😢  I'm sorry you ran into this. We also have similar struggles dealing with developing on OSX and the CI runners using linux. Thank you for pushing past that 🙏

estherk15 commented on website/cue/reference/components/sources/base/http.cue:
```diff
@@ -345,6 +345,20 @@ base: components: sources: http: configuration: {
 	}
+	host_key: {
+		description: """
+			Overrides the name of the log field used to add the remote ip to each event.
```
```suggestion
			Overrides the name of the log field used to add the remote IP to each event.
```

estherk15 commented on website/cue/reference/components/sources/base/http_server.cue:
```diff
@@ -345,6 +345,20 @@ base: components: sources: http_server: configuration: {
 	}
+	host_key: {
+		description: """
+			Overrides the name of the log field used to add the remote ip to each event.
```
```suggestion
			Overrides the name of the log field used to add the remote IP to each event.
```

estherk15 reviewed (approved):
Non-blocking suggestion.

neuronull commented:
👋  hey @NishantJoshi00  

I ran CI on this PR and it is looking great.

I think the suggestion from @estherk15  is worth making.

One thing I forgot but CI found- this change is an enhancement which creates a user experience change, and should have a changelog fragment to coincide.

It's a fairly straightforward process, described here: hhttps://github.com/vectordotdev/vector/blob/master/changelog.d/README.md


Here is an example, note you need to specify `authors` at the end of the file, and a link to your GH profile will be rendered in the release notes as the contributor to this change.
 https://github.com/vectordotdev/vector/blob/master/changelog.d/19639_unit_test_vrl_metadata.enhancement.md?plain=1

jszwedko commented on src/sources/http_server.rs:
```diff
@@ -484,6 +510,16 @@ impl HttpSource for SimpleHttpSource {
+                            log,
+                            self.host_key.path.as_ref().map(LegacyKey::Overwrite),
+                            path!("host"),
+                            addr.to_string(),
```
I believe this will include the port number (like 10.12.23.32:12739), is that what you expected?

jszwedko replied:
If so, could we update the docs to say they that the value will include the port and not just the IP address?

jszwedko replied:
(this also highlights a lack of tests in this PR 😄 Could we add one?)

NishantJoshi00 replied:
Hey @jszwedko, yes I am aware that this would also include the port number, but I wasn't convinced that we should have it, or exclude it. What do you suggest? If yes, how can this be helpful in observability setup?

NishantJoshi00 replied:
I feel, we should only have host IP, according to the requirements discussed this far.

jszwedko replied:
I agree, I believe that is the precedent we have for our sources (to only include IP address for `host`).

NishantJoshi00 commented:
Hey @neuronull , can you please verify if the changelog entry is up to the mark, specified by the requirements

neuronull commented on changelog.d/4221_host_key_http_source.enhancement.md:
```diff
@@ -0,0 +1,3 @@
+Add support to display the remote IP that sent log data to Vector via the `http` source.
```
I would extend this a little, to note the new setting, something like


```suggestion
Added an event field to the `http` source which provides the remote IP that sent log data to Vector.

The event field key is configurable through a new configuration setting `host_key`, which defaults to the global host key.
```

neuronull commented:
>  Hey @neuronull, my name is failing the spell check 😢

Ohh sorry about that, I believe we fixed this on master so you should just need to pull master into your branch. I attempted this but was not able to checkout your PR for some reason.

NishantJoshi00 commented:
Hey @neuronull, I changed inline comments to make them aligned to the changes that @estherk15 suggested.

neuronull commented:
Huh. Interesting, this change has flagged a performance regression in two test cases that exercise this component.

I'm wondering if its this bit of code


```Rust
                    .map_request(move |mut request: hyper::Request<_>| {
                        request.extensions_mut().insert(PeerAddr(remote_addr));
```

NishantJoshi00 commented:
Yes @neuronull, that's highly likely the case. 

I tried to look around, nothing else seems to be introducing this cost. Neither `SocketAddr` nor the middleware that was introduced. It has either something to do with adding an extension to the request or the closure capturing the remote IP on every new connection (this shouldn't introduced such regression).

It might just be introduced by extension!

NishantJoshi00 commented:
Hey @neuronull, can you tell me the command to do this benchmarking and regression testing locally?

Is it "make bench"?

neuronull commented:
👋  Hey @NishantJoshi00 , thanks for volunteering to investigate that.

Unfortunately there isn't a "pure" way to get the exact same experiment report locally, as we do in CI. There is some internal infrastructure being leveraged there which isn't exposed (even to the Vector team).

However, our `regression` tests (see regression/) , are using a tool called lading. This is in a public repo and can be run locally. So I believe you could run the failing experiments ("cases" in our repo) locally, using that tool:

https://github.com/DataDog/lading

jszwedko commented:
> what is the base value around which this delta was calculated?

The baseline is determined from the merge base of the PR and `master`: https://github.com/vectordotdev/vector/blob/2f1c7850fbc039a894f51b844e919adf2fdc925d/.github/workflows/regression.yml#L151 . To determine performance differences in a PR, CI will build both the baseline and PR, run both `n` times for each experiment (I think 20 if I remember correctly), and compare the averages.

NishantJoshi00 commented:
Hey @neuronull @jszwedko. So, I have a hypothesis, let me know if it makes sense.
Currently we are passing the entire `SocketAddr` in the extensions section of the `Service` handlers.

Instead, what if we pass a reference to it, in this scenerio we cannot solve this with `&` as this will push us into a lifetime rabbit-hole, what if instead we pass a `Arc`. In this situation, instead of passing the entire `SocketAddr` we are just passing the reference to it.

| Type | Size (bytes) |
| ---    |   ---              |
| SocketAddr | 32 | 
| Arc | 8 | 

This will reduce the size of the data that is being generated, copied and sent during every connection, which might optimize the perf.

I'wd like to know your thoughts on this hypothesis, and test this change too.

neuronull commented:
> I'wd like to know your thoughts on this hypothesis

Looks good to me, I will re-trigger CI

jszwedko commented:
@NishantJoshi00 I think it's worth a shot if you want to try the `Arc` approach. It looks like we are still seeing a regression.

jszwedko commented:
An alternative here would be to make the field addition opt-in via configuration.

NishantJoshi00 commented:
Hey @jszwedko, I apologies for the delay. I didn't get time to open and work on this PR for a few weeks. But, I am back and I can try and resolve this PR.

> An alternative here would be to make the field addition opt-in via configuration.

By this, you mean compile time feature? because this is a framework level change and adding a runtime flag on this seems to be a challenge. My guess is that regression cost is added not because we are extracting the `remote_addr` but because we are plugging it back in using `Extension`.

I was thinking around the regression.

- Does the regression consider situation around communicating over a open connection? Or only regresses over data transferred over new connections?
   As by theory, the performance shouldn't be affected if data is being sent over an open connection.
- The advantage of adding `Arc` wasn't as significant and `SocketAddr` already has `Copy` should we remove that?
- `http_text_to_http_json` what does this regression indicate?

Can you elaborate more on how I can make this a opt-in feature?

NishantJoshi00 commented:
Hey @jszwedko,
> If that is empty, the host key is not added. I think we would just default to empty and let users opt into it.

So, I am giving this a shot, my assumption is that the cost is added not because of adding a extension to the service, but by performing logical operations while creating connections, but if the operations is as light as a `jmp` then, the effect should be minimal and after a bit a reconsideration this seems worth a shot. can you run the regression test for this?

jszwedko commented:
/ci-run-regression

jszwedko commented on changelog.d/4221_host_key_http_source.enhancement.md:
```diff
@@ -0,0 +1,5 @@
+Added an event field to the `http` source which provides the remote IP that sent log data to Vector.
+
+The event field key is configurable through a new configuration setting `host_key`, which defaults to the global host key.
```
```suggestion
The `http_server` source can now optionally annotate events with the remote IP by setting the `host_key` option.
```

Shortening this up a bit.

jszwedko commented on src/sources/http_server.rs:
```diff
@@ -23,8 +23,13 @@ use vector_lib::{
+        SourceConfig,
+        SourceContext,
         SourceOutput,
+        // log_schema
```
```suggestion
```

jszwedko commented on src/sources/http_server.rs:
```diff
@@ -131,6 +136,12 @@ pub struct SimpleHttpConfig {
     path_key: OptionalValuePath,
 
+    /// Overrides the name of the log field used to add the remote IP to each event. By default, the remote IP is not added.
+    ///
```
```suggestion
```

jszwedko commented on src/sources/http_server.rs:
```diff
@@ -131,6 +136,12 @@ pub struct SimpleHttpConfig {
     #[configurable(metadata(docs::examples = "vector_http_path"))]
     path_key: OptionalValuePath,
 
+    /// Overrides the name of the log field used to add the remote IP to each event. By default, the remote IP is not added.
```
```suggestion
    /// If set, the name of the log field used to add the remote IP to each event
```

jszwedko commented on src/sources/http_server.rs:
```diff
@@ -284,6 +303,11 @@ fn default_path_key() -> OptionalValuePath {
 }
 
+fn default_host_key() -> OptionalValuePath {
+    // log_schema().host_key().cloned().into()
```
```suggestion
```

jszwedko commented on src/sources/util/http/prelude.rs:
```diff
@@ -184,15 +191,26 @@ pub trait HttpSource: Clone + Send + Sync + 'static {
+                        if let Some(remote_addr_inner) = remote_addr_ref.as_ref() {
+                            request
+                                .extensions_mut()
+                                .insert(PeerAddr::new(Arc::clone(&remote_addr_inner)));
```
Clippy flagged this.

jszwedko commented on src/sources/util/http/prelude.rs:
```diff
@@ -184,15 +191,26 @@ pub trait HttpSource: Clone + Send + Sync + 'static {
             let span = Span::current();
             let make_svc = make_service_fn(move |conn: &MaybeTlsIncomingStream<TcpStream>| {
+                let remote_addr = conn.peer_addr();
+                let remote_addr_ref = enable_source_ip.then(|| Arc::new(remote_addr));
```
Do you still think the `Arc` is helpful? I know you mentioned before that you didn't think it made a difference. We could drop it if so.

NishantJoshi00 replied:
Ya, we can explore this idea.

NishantJoshi00 replied:
yes, seems like removing `Arc` definitely be helpful.

<img width="699" alt="image" src="https://github.com/vectordotdev/vector/assets/51093026/1a961f3b-b4bb-48ae-b33a-67d74843d411">

This was the difference between using `Copy` and `Arc::clone` which I observed

jszwedko reviewed (commented):
Awesome, this looks like it did the trick @NishantJoshi00 ! I think it just needs some clean-up and then we can merge it. Thanks for your persistence on this one!

jszwedko commented:
/ci-run-regression

jszwedko reviewed (approved):
Thanks again for your work on this!

NishantJoshi00 commented:
Seems like a test is failing:
```diff
---- sources::http_server::tests::output_schema_definition_legacy_namespace stdout ----
thread 'sources::http_server::tests::output_schema_definition_legacy_namespace' panicked at src/sources/http_server.rs:1555:9:
assertion failed: `(left == right)`'
  left: `"Some(Definition { event_kind: Kind { bytes: None, integer: None, float: None, boolean: None, timestamp: None, regex: None, null: None, undefined: None, array: None, object: Some(Collection { known: {F..."` (truncated)
 right: `"Some(Definition { event_kind: Kind { bytes: None, integer: None, float: None, boolean: None, timestamp: None, regex: None, null: None, undefined: None, array: None, object: Some(Collection { known: {F..."` (truncated)

Differences (-left|+right):
                 Collection {
                     known: {
                         Field(
                             KeyString(
+                                "host",
+                            ),
+                        ): Kind {
+                            bytes: Some(
+                                (),
+                            ),
+                            integer: None,
+                            float: None,
+                            boolean: None,
+                            timestamp: None,
+                            regex: None,
+                            null: None,
+                            undefined: None,
+                            array: None,
+                            object: None,
+                        },
+                        Field(
+                            KeyString(
                                 "message",
                             ),
                         ): Kind {
                             bytes: Some(
```

I just verified both the side of the `assert` statement, as the failure suggests the left side doesn't contain the information about the `host` key. But, I just checked the `schema_definition` function and it does seem to have the definition about `host`. Did I miss something, somewhere else?

jszwedko commented:
Oh yes. I think the issue is that you need to update the `SimpleHttpConfig::schema_definition` function to add the schema definition for the new field. I think it should be `Kind::bytes().or_undefined()` because the field isn't always there (which means you should update the test too).

NishantJoshi00 commented:
Though it seems that like `output_schema_definition_vector_namespace` this test passed. I am unclear on the difference between the Legacy NS and Vector NS

jszwedko commented:
> Though it seems that like `output_schema_definition_vector_namespace` this test passed. I am unclear on the difference between the Legacy NS and Vector NS

The Vector one tests with [log namespacing](https://vector.dev/blog/log-namespacing/) enabled.

I think you still need to update the `SimpleHttpConfig::schema_definition` function. I'm not seeing any modifications there.

NishantJoshi00 commented:
Hey @jszwedko, I have changed it in `SimpleHttpConfig::schema_definition`

```rust
            .with_source_metadata(
                SimpleHttpConfig::NAME,
                self.host_key.path.clone().map(LegacyKey::Overwrite),
                &owned_value_path!("host"),
                Kind::bytes().or_undefined(),
                None,
            )
```

I have fixed fmt check

jszwedko commented:
Ah, gotcha, I somehow missed that. Let me rerun CI.
